server/src/main/java/org/opensearch/index/ShardIndexingPressure.java

```diff
@@ -0,0 +1,360 @@
+        if (forceExecution == false) {
+            boolean nodeLevelLimitBreached = memoryManager.isCoordinatingNodeLimitBreached(tracker, nodeTotalBytes);
+            if (nodeLevelLimitBreached == false) {
+                shardLevelLimitBreached = memoryManager.isCoordinatingShardLimitBreached(tracker, nodeTotalBytes, requestStartTime);
+            }
+
+            if (shouldRejectRequest(nodeLevelLimitBreached, shardLevelLimitBreached)) {
+                long nodeBytesWithoutOperation = nodeCombinedBytes - bytes;
+                long nodeTotalBytesWithoutOperation = nodeTotalBytes - bytes;
+                long shardBytesWithoutOperation = shardCombinedBytes - bytes;
+
+                currentCombinedCoordinatingAndPrimaryBytes.addAndGet(-bytes);
+                coordinatingRejections.getAndIncrement();
+                tracker.getCommonOperationTracker().incrementCurrentCombinedCoordinatingAndPrimaryBytes(-bytes);
+                tracker.getCoordinatingOperationTracker().getRejectionTracker().incrementTotalRejections();
+
+                throw new OpenSearchRejectedExecutionException("rejected execution of coordinating operation [" +
+                    "shard_detail=[" + shardId.getIndexName() + "][" + shardId.id() + "][C], " +
+                    "shard_coordinating_and_primary_bytes=" + shardBytesWithoutOperation + ", " +
+                    "shard_operation_bytes=" + bytes + ", " +
+                    "shard_max_coordinating_and_primary_bytes=" + tracker.getPrimaryAndCoordinatingLimits() + "] OR [" +
+                    "node_coordinating_and_primary_bytes=" + nodeBytesWithoutOperation + ", " +
+                    "node_replica_bytes=" + nodeReplicaBytes + ", " +
+                    "node_all_bytes=" + nodeTotalBytesWithoutOperation + ", " +
+                    "node_operation_bytes=" + bytes + ", " +
+                    "node_max_coordinating_and_primary_bytes=" + primaryAndCoordinatingLimits + "]", false);
+            }
+        }
+        currentCoordinatingBytes.addAndGet(bytes);
+        totalCombinedCoordinatingAndPrimaryBytes.addAndGet(bytes);
+        totalCoordinatingBytes.addAndGet(bytes);
+
+        StatsTracker statsTracker = tracker.getCoordinatingOperationTracker().getStatsTracker();
+        statsTracker.incrementCurrentBytes(bytes);
+        markShardOperationStarted(statsTracker, tracker.getCoordinatingOperationTracker().getPerformanceTracker());
+        boolean isShadowModeBreach = shardLevelLimitBreached;
+
+        return () -> {
+            currentCombinedCoordinatingAndPrimaryBytes.addAndGet(-bytes);
+            currentCoordinatingBytes.addAndGet(-bytes);
+            markShardOperationComplete(bytes, requestStartTime, isShadowModeBreach, tracker.getCoordinatingOperationTracker(),
+                tracker.getCommonOperationTracker());
+            memoryManager.decreaseShardPrimaryAndCoordinatingLimits(tracker);
+            tryReleaseTracker(tracker);
```

Comment:
This is avoidable per request when all thats needed on the tracker is 
```
final Map<ShardId, ShardIndexingPressureTracker > shardIndexingPressureHotStore =
        Collections.synchronizedMap(new LinkedHashMap<ShardId, ShardIndexingPressureTracker>(100, .75F, true) {
            @Override
            protected boolean removeEldestEntry(Map.Entry eldest) {
                return size() > 100;
            }
        });
```

Reply:
This will limit the hot shards i.e. shards tracking inflight requests to 100, and may remove the ones currently in use. However with the approach we want to allows hot shards to grow based on the actual traffic, and purge only the ones in cold store. 
We have an item to evaluate and refactor shard store cleanup strategy as a follow up going forward #478.

Reply:
```
final Map<ShardId, ShardIndexingPressureTracker > shardIndexingPressureColdStore =
        Collections.synchronizedMap(new LinkedHashMap<ShardId, ShardIndexingPressureTracker>(100, .75F, true) {
            @Override
            protected boolean removeEldestEntry(Map.Entry eldest) {
                return size() > 100;
            }
        });
```
Does this work now?

Reply:
Yes this will help with the eviction strategy for cold store, however, we will still need to move the trackers from hot to cold store conditionally, which is happening here. Also, this will change the cold store from hash map to linked hash map. I will evaluate and revisit the cold store cleanup strategy as a follow up item listed in #478.

Reply:
This is a very pessimistic unnecessary operation per request since a small young GC pause can cause trackers moving between hot and cold.

Reply:
> This is a very pessimistic unnecessary operation per request since a small young GC pause can cause trackers moving between hot and cold.

Just to be more clear, this movement is not per request but only when all the requests for a shards drains out from the node, and there were no new incoming requests. This should not cause any churns in high tps scenarios, with continous or concurrent workload, which seems to be a concern here. 
Also, when the tracker  movements happens it is a not a deep copy, and no new tracker objects are created, but only references are copied, so overhead is minimal.

>IMHO we are just complicating things here when a single tracker would have been sufficient and could have been periodically cleared by a background sweep rather than per request checks to move trackers back and forth.

The two phased lifecycle of tracker (hot and cold) has actually helped reduce churns and reduce GC, as found in the allocation profiling data. Will be sharing the performance results post we complete the PR for plumbing the logic. This has also simplified from implementing any background sweep logic, for now. 

>Please think about optimizing this and open a dedicated issue highlighting this problem so that it doesn't get missed.

I am open to suggestions and inputs to optimise this further. Given it is not critical, will like to push it behind as a follow up. Currently all related PRs and future pending items are being tracked at meta #478 level, and I have added an open item entry for the same here, so that this is not missed. Let me know if you still think we need a dedicated issue for this.

Reply:
Here is what we should think about, listing few concerns with tracking and cleaning
1. Current implementation delegates this responsibility externally, which could well be placed within the tracker to manage the lifecycle. It might be hard to track clean up and maintain overtime
2. Just becoz the current implementation is doing some optimization doesn't mean other options discussed aren't performant, you should consider evaluating other proposals as well
3. Client side batching delays, N/W delays, connection establishment all of this can lead to premature `drain` and aggressive clean up which is a sub-optimal approach of clean up. Most of which is something which the server cannot control and hence shouldn't control. So it would impact `high tps` traffic as well
4. The clean up doesn't have to be client side facing, we need to understand the additional overhead. In contrast background cleanup are relatively cheaper and can take better decision atleast better than the weak condition

```
tracker.getCommonOperationTracker().getCurrentCombinedCoordinatingAndPrimaryBytes() == 0 &&
                tracker.getReplicaOperationTracker().getStatsTracker().getCurrentBytes() == 0
```

All of these are suggestions its fine even if you don't want to take up

Reply:
While I agree to most of it, there are few important considerations to make too, and hence was suggesting to de-couple this from the current PR. Captured the high level thoughts here in a dedicated issue #1033, in addition to item in #478 we previously had, to evaluate and follow up. Let me know if you think it's something should not be decoupled or is critical.